alica_engine/src/engine/Assignment.cpp

```diff
@@ -1,532 +1,448 @@
+            std::copy(_successData.getRaw()[i].begin(), _successData.getRaw()[i].end(), std::back_inserter(o_agents));
+            return;
         }
     }
+}
 
-    auto robotsSucc = this->epSucMapping->getRobots(ep);
-    if (robotsSucc != nullptr) {
-        for (auto& robot : *robotsSucc) {
-            ret->push_back(robot);
+void Assignment::getAgentsInState(const State* s, AgentGrp& o_agents) const
+{
+    const EntryPoint* ep = s->getEntryPoint();
+    for (AgentStatePair asp : _assignmentData[ep->getIndex()]) {
+        if (asp.second == s) {
+            o_agents.push_back(asp.first);
         }
     }
-
-    return ret;
 }
 
-/**
- * The robots that are currently working on or already succeeded in a specific task, referred to by an EntryPoint.
- * Each robot only occurs once.
- * @param ep An entrypoint
- * @return a shared_ptr of a list of int
- */
-shared_ptr<list<const supplementary::AgentID*>> Assignment::getUniqueRobotsWorkingAndFinished(const EntryPoint* ep)
-{
-    auto ret = std::make_shared<list<const supplementary::AgentID*>>(list<const supplementary::AgentID*>());
-    // if (this->plan->getEntryPoints().find(ep->getId()) != this->plan->getEntryPoints().end())
-    {
-        auto robots = this->epRobotsMapping->getRobotsByEp(ep);
-        for (int i = 0; i < static_cast<int>(robots->size()); ++i) {
-            ret->push_back(robots->at(i));
-        }
-        for (auto r : (*this->epSucMapping->getRobots(ep))) {
-            if (std::find_if(ret->begin(), ret->end(), [&r](const supplementary::AgentID* id) { return *r == *id; }) == ret->end()) {
-                ret->push_back(r);
-            }
-        }
-    }
-    return ret;
+AssignmentView Assignment::getAgentsWorking(const EntryPoint* ep) const
+{
+    return AssignmentView(this, ep->getIndex());
 }
 
-/**
- * The robots that are currently working on or already succeeded in a specific task, referred to by an EntryPoint Id.
- * @param epid EntryPoint id
- * @return a shared_ptr of a list of int
- */
-std::shared_ptr<list<const supplementary::AgentID*>> Assignment::getRobotsWorkingAndFinished(int64_t epid)
+AssignmentView Assignment::getAgentsWorking(int idx) const
 {
-    auto ret = std::make_shared<std::list<const supplementary::AgentID*>>();
-    auto robots = this->epRobotsMapping->getRobotsByEpId(epid);
-    if (robots != nullptr) {
-        for (int i = 0; i < static_cast<int>(robots->size()); i++) {
-            ret->push_back(robots->at(i));
-        }
-    }
+    return AssignmentView(this, idx);
+}
 
-    auto robotsSucc = this->epSucMapping->getRobotsById(epid);
-    if (robotsSucc != nullptr) {
-        for (auto& robot : *robotsSucc) {
-            ret->push_back(robot);
-        }
-    }
+AssignmentView Assignment::getAgentsWorking(int64_t epId) const
+{
+    const EntryPoint* ep = _plan->getEntryPointByID(epId);
+    return ep ? AssignmentView(this, ep->getIndex()) : AssignmentView();
```

Comment:
Since entrypoint ids are static, and because this is potentially called quite often, I might just assume that the id passed into this function MUST exist, so we can get rid of the unnecessary check and not have the user do further sanity checking once they receive the empty AssignmentView. I think if an invalid entrypoint id is passed in here, we should fail.

Then we might not need the default constructor for AssignmentView, and this function can simply be: `return getAgentsWorking(_plan->getEntryPointByID(epId));`

Reply:
The id might exist, but maybe not in the plan you ask for it.
The default constructor is useful to generate empty views whenever invalid cases are asked for (not only here).

Reply:
I guess my question is, why not crash if the user provides an invalid id? If the user did something like that, I don't think the program should continue running, otherwise it might look like everything the user specified was correct, when it wasn't. Since all of these ids and the plan structures are generated by the plan designer before this code is even compiled, one optimization we could make is to use static_asserts to verify correct usage of ids.